au/code/au/utility/test/probable_primes_test.cc

```diff
@@ -200,6 +200,115 @@ TEST(MillerRabin, SupportsConstexpr) {
+    const auto primes = first_n_primes<3'000u>();
+    for (const auto &p : primes) {
+        if (p > 2u) {
+            EXPECT_THAT(strong_lucas(p), Eq(PrimeResult::PROBABLY_PRIME)) << p;
+        }
+    }
+}
+
+TEST(StrongLucas, GetsFooledByKnownPseudoprimes) {
+    for (const auto &p : strong_lucas_pseudoprimes()) {
+        ASSERT_EQ(miller_rabin(2u, p), PrimeResult::COMPOSITE) << p;
+        EXPECT_THAT(strong_lucas(p), Eq(PrimeResult::PROBABLY_PRIME)) << p;
+    }
+}
+
+TEST(StrongLucas, OddNumberIsProbablyPrimeIffPrimeOrPseudoprime) {
+    const auto primes = first_n_primes<3'000u>();
+    const auto pseudoprimes = strong_lucas_pseudoprimes();
+
+    // Make sure that we are both _into the regime_ of the pseudoprimes, and that we aren't off the
+    // end of it.
+    ASSERT_THAT(primes.back(), AllOf(Gt(pseudoprimes.front()), Lt(pseudoprimes.back())));
+
+    std::size_t i_prime = 1u;  // Skip 2; we're only checking odd numbers.
+    std::size_t i_pseudoprime = 0u;
+    for (uint64_t n = primes[i_prime]; i_prime < primes.size(); n += 2u) {
+        const auto is_prime = (n == primes[i_prime]);
+        if (is_prime) {
+            ++i_prime;
+        }
+
+        const auto is_pseudoprime = (n == pseudoprimes[i_pseudoprime]);
+        if (is_pseudoprime) {
+            ++i_pseudoprime;
+        }
+
+        const auto expected =
+            (is_prime || is_pseudoprime) ? PrimeResult::PROBABLY_PRIME : PrimeResult::COMPOSITE;
+        EXPECT_THAT(strong_lucas(n), Eq(expected)) << "n = " << n;
+    }
+}
+
+TEST(BailliePSW, BadInputForLessThanTwo) {
+    EXPECT_THAT(baillie_psw(0u), Eq(PrimeResult::BAD_INPUT));
+    EXPECT_THAT(baillie_psw(1u), Eq(PrimeResult::BAD_INPUT));
+}
+
+TEST(BailliePSW, TwoIsPrime) { EXPECT_THAT(baillie_psw(2u), Eq(PrimeResult::PROBABLY_PRIME)); }
+
+TEST(BailliePSW, CorrectlyIdentifiesAllOddNumbersUpToTheFirstThousandPrimes) {
+    const auto first_10k_primes = first_n_primes<10'000u>();
+
+    std::size_t i_prime = 1u;  // Skip "prime 0" (a.k.a. "2").
+    for (uint64_t i = 3u; i_prime < first_10k_primes.size(); i += 2u) {
+        const bool is_prime = (i == first_10k_primes[i_prime]);
+        if (is_prime) {
+            ++i_prime;
+        }
+        const auto expected = is_prime ? PrimeResult::PROBABLY_PRIME : PrimeResult::COMPOSITE;
+        EXPECT_THAT(baillie_psw(i), Eq(expected)) << "i = " << i;
+    }
+}
+
+TEST(BailliePSW, IdentifiesPerfectSquareAsComposite) {
+    // (1093 ^ 2 = 1,194,649) is the smallest strong pseudoprime to base 2 that is a perfect square.
+    constexpr auto n = 1093u * 1093u;
+    ASSERT_THAT(miller_rabin(2u, n), Eq(PrimeResult::PROBABLY_PRIME));
+    EXPECT_THAT(baillie_psw(n), Eq(PrimeResult::COMPOSITE));
+}
+
+TEST(BailliePSW, HandlesVeryLargePrimes) {
+    for (const auto &p : {
+             225'653'407'801ull,
+             334'524'384'739ull,
+             9'007'199'254'740'881ull,
+             18'446'744'073'709'551'557ull,
```

Comment:
Rather than relying on `ull` to be equivalent to a `uint64_t`;

```suggestion
             uint64_t{225'653'407'801},
             uint64_t{334'524'384'739},
             uint64_t{9'007'199'254'740'881},
             uint64_t{18'446'744'073'709'551'557},
```

Reply:
+1. Luckily, `ull` is produces literals of a size of 8 bytes on clang, gcc, and MSVC. It's not guaranteed. Literals with `ul` are 8 bytes on clang and gcc, but 4 bytes on MSVC. [godbolt](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsQAJnL0WTUKgCkBgEJXr5FY1RECQ6s0wBhdPQCubJkMADnIPABkCJmwAOX8AI2xSEAA2cgAHdCViVyZvPwDg9MzsoQio2LYEpNTHbGcckSIWUiI8/0CDENr6oUbmojKY%2BMSUhyaWtoLOsf7BiqqUgEoHdF9SVE4uKwBmSNQ/HABqC23PVCUiQmET3AstAEEdvYPsY9PXC9Jsdhu7x4fIkRDmwWJEIItjgB2WwPQ5ww4XTAgNCrIEnTzohEEABe2HQNAgvkByWkAH0iBD0ZiwFsAKyeJg0k4w%2B7whGXZEYXxo06YrK4/EQHi%2BeiU3mnQ5M%2BmMrbbFlsxGc1FvDES/l4gnC%2BiilXUukMplyv5/NlfIhrJiHLTM42QgAiXGW9G4tP4gS4OnI6G4GJsdgRq3Wrx2Bn4RG0juWSG%2BOCS4PIAGsQLStAA6SGQ2nbHhBaQGbYGZKQ5LbWnGbjSfhsQxacjuz3erj8JQgWvhj2O8hwWAoDBsNIMRKUah9geMJIAN1QaTSpInPAAnKTpFpSQA1EQ8SGknir9RBVJ0ehERItiBxCPkOKRZoAT24fCvN9It4A8nFdHV24%2B%2BxxhK%2BmHoe8O3IHA4l8YBPAkegW14fgcBBMxJBAwgvnqCdsFgz1sHUOpuU2R9AWwZ0QPoAg4lIO9vBwS8iFIAhqzg8gMNIOJMmwO1sEQ4AyLMCNlhoExgCUNcCGwAB3V80mYB9%2BEEYQxAkTgZDkYRlDUTQQP0HhjFMcw7DsYxyJbSBlnQNIXCEWD%2BHQFj6JwEz426SzAggDxJkCHTwkiIZKhGHSMiyFyPKkIogpyOZhiSHTnIacZWh8dpQti3p4sivzopmCZEoKGK0p8%2BZ/OWJRAw2QwnRdN1L0bQ592SYElCnQ55wXVMV0ODct1TXdDggfBiDIY58wMRYw34qMYxGeMSMrchqx4ZJUwLZIsy0ZIeAMAxpGkBc62q7hm1bch2x0ZZu2QEAcLwkgKCoCBmmE1RTGIkQkHQcT3R/dB%2BwYFgXKeqJ6Fe97634UdB0nadZxa5dV067dd1JOryHB8donYTYwe%2BsdElfblgY%2By8rtQe5SGEg7Qlw1BGiuCn5NEcRJBU%2Bn1I0S99CMEw%2BMsP0bCMuJHLMiycmsr07IIBz4GWUhfGEBjsAAFXQHxBZWNYyqMRFIgBl63sJpi6OwTha3Eyi0lkiquFdPaQMbXAqe5Qa6oapqWrarQOs3SFuo9iBfVsPnDn6m6hoLUbjvG8suFm6t81TIttmkZIF22IIM1pWkdNBr0KZbNt%2BK7RAIF7bGIeHCBUZGYAZCMI8T1IM8LxA69WBfWSn1bt8Py/dvf2YIgAKAy8wIgqDtVgx8EL05DPVQr8CAwrD%2BGJ/D26IkjPTIiiqKwTZPTohj25YtiVE47jeNADsBKEkSxMk6TPrk%2BRFKZ2QWZUNmtJAbZdO5gy%2Ba3qrcyLlRa2USPZTCUsHDEXnm4NyTAvA5U8qEBB6UFgBWKMFJBoVAolCYGg/y0CnAuT6NlfIyCUpMFIQMAqUVkrxRCnlWYtCMpSGKqVTgI0o7W2zjVOqABaEkhxgCoFQIcGQqYDC9WDoNEM4cTqRijjHGsNsGy5yOgos6RcUArxuuXB6SgdZAz1o/FGpcTD/WesYkGl5K5JGrnmMxP00YYwpnYvGogTFEypqTcmTZKZ1BppEOmz9GbKTfvIVmmlPT6B/lzfSvN7CAKgcAkWFMwGkAgarGWcsOBKxVlAkq6tOFjCuEYgmpjDbG3IKbFg5s4KWx4ftLg9trpOwPII6QwjRHiOkJI3q/tDJB0ICHORY0r6W2UQYWsvD1H5wmYmZMaZ07ZlzMNIsJYywkW2FVW2FNNGF3gOdYuKJnFDjunYtA8TFy7gEAweujdLwtzvO3Z5L53yfmcL3b6f4B6AWArPbA4FILQQnvBLi0897wQIGhFwi9vHXQIvwdel4t6URfNRJFx16KMUfMfdiZ9p6REvqdAQN9RISSkjJJi9MX7hNUooD%2B0S9CGF/gkgOSTjIpOFlZdJ4tJamSIbA1y7lsFeVQSw9BYU8GMOlS5AhmVKHUNlUq/K5Q6FMLIUlTVND1WsJ4Ow4p5VuG7LUVwWqHShH7D0uI1qPBUy%2BxkaQUOI1xmnWWEmaQQRUxBG2JCLQm1tgrS0AuIItJUgzSrMmPpyQghrRJAuBaQQw2yFmf4vOEcr6HJ7CAVEaRuTlzsejDg3ABFWvieI2kDqHXLxGWQCWrLaVhKkBEtSTL2astqfUvgjTTU2W4B4/NQJ8QWuSJ0w41qzC2u6g6gZ5jEguvkZHaMLBYzUEmVGng9rQ3JC0FoLcO1IRBA2hmVR/b00aILscnRDs9F3QMeUrxTFUZ/RyI%2BmxIFLmTurom2sRbXH%2BPcfjJ92EfFk1FsTIJ7on4KWbczSJ7av46XiWgf%2BnKBbcpAdwfhr5tiHH4YiE4doeYcvzDZflkDBU5JcHk5WMFCkcNZVrJg779aPiqUdLtFsTVnpzi0297Sx3lptYuGdvtBmBydYut1ijI1zW/tsJa2xlMqdUyp3jjZDrzPdYsnakitqxp4LSYskICypyjjsjT%2BzI4kVDFZ/xmjlgsSyG4aQQA%3D%3D)

Reply:
Literals generally are a funny and subtle thing.  The type of an integral literal will be the default integral type (usually `int` or `uint`), _unless_ the value won't fit in that type, in which case, the type will be the smallest type where it _will_ fit, being signed for no suffix and unsigned for `u`.  (I don't remember where I learned this.)

The problem this causes is that if you use a simple braced list like this, and just use a `u` suffix for everything, the compiler will _deduce inconsistent types_ if some of those literals can fit in a `uint`, and some can't.  Therefore, I decided to apply a literal suffix that would cover all of them.

But, the feedback in this thread shows why that's not a great idea.  Explicit `uint64_t` seems much more readable.  (The other alternative would be to make an explicit `std::vector<uint64_t>`, which would make mixing-and-matching OK.)